docs/check.go

```diff
@@ -0,0 +1,294 @@
+	// check that script is being run from repo root
+	const docsDir, configYml = "docs", "_config.yml"
+	repoRoot, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("could not get current working directory: %v", err)
+	}
+	docsPath, err := filepath.Abs(filepath.Join(repoRoot, docsDir))
+	if err != nil {
+		return fmt.Errorf("could not get absolute path of %v: %v", filepath.Join(repoRoot, docsDir), err)
+	}
+	configPath := filepath.Join(docsPath, configYml)
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		return fmt.Errorf("expected to find %s in %s; script should be executed from repository root", configYml, docsDir)
+	}
+
+	// parse category names from config file
+	categoryNames, err := parseCategoryNames(configPath)
+	if err != nil {
+		return fmt.Errorf("error parsing category names: %v", err)
+	}
+
+	ctx := &checkContext{
+		categoryNames:  categoryNames,
+		docsPath:       docsPath,
+		knownFiles:     StringSet{},
+		markdownLinks:  map[string][]*relativeLink{},
+		markdownErrors: map[string][]string{},
+	}
+
+	// scan through markdown files
+	for _, langDir := range []string{"en", "cn"} {
+		if err := filepath.Walk(filepath.Join(docsPath, langDir),
+			func(p string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if info.IsDir() {
+					return nil
+				}
+				if strings.HasSuffix(info.Name(), ".md") {
+					if err := checkFile(p, ctx); err != nil {
+						return err
+					}
+				}
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error traversing through md files in %v: %v", filepath.Join(docsPath, langDir), err)
+		}
+	}
+	// scan through img and resources directories to update known files
+	for _, dir := range []string{"img", "resources"} {
+		if err := filepath.Walk(filepath.Join(docsPath, dir),
+			func(p string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if info.IsDir() {
+					return nil
+				}
+				ctx.knownFiles.Add(strings.TrimPrefix(p, docsPath))
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error traversing through files in %v: %v", filepath.Join(docsPath, dir), err)
+		}
+	}
+
+	ctx.checkLinks()
+
+	if len(ctx.markdownErrors) > 0 {
+		errLines := []string{"Errors found in documentation markdown"}
+		for f, errs := range ctx.markdownErrors {
+			errLines = append(errLines, fmt.Sprintf("  %v:", strings.TrimPrefix(f, repoRoot)))
+			for _, err := range errs {
+				errLines = append(errLines, fmt.Sprintf("    %s", err))
+			}
+		}
+		return fmt.Errorf("%v", strings.Join(errLines, "\n"))
+	}
+
+	return nil
+}
+
+// parseCategoryNames parses the given config file for the list of category names
+func parseCategoryNames(configPath string) (StringSet, error) {
+	f, err := os.Open(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file at %v: %v", configPath, err)
+	}
+	defer f.Close()
+
+	const categoryListLine, categoryPrefix = "categoryList:", "  - "
+	categoryNames := StringSet{}
```

Comment:
I think we should make a struct and parse YAML instead of using this string manipulation

Reply:
parsing yaml -> vendoring code -> need to set GOPATH to execute
the generate tarball code needs to do the lattermost because it consists of several files, but i've been of the opinion of keeping go code in this repo as scripts. i want to avoid vendoring code unless we're completely on board with integrating all of the go code here into its own module as opposed to one-off scripts

Reply:
I would prefer to remove the go code if possible. Since the docs use jekyll, it would probably be easier (assuming you know ruby) to write a verification program that can run as a [plugin](https://jekyllrb.com/docs/plugins/).

Reply:
never written ruby...